HW2/P5/physics.pyx

```diff
@@ -50,37 +50,49 @@ cdef inline void collide(FLOAT *x1, FLOAT *v1,
         v1[dim] -= change_v1[dim]
         v2[dim] += change_v1[dim]  # conservation of momentum
 
-cdef void sub_update(FLOAT[:, ::1] XY,
+cdef void sub_update(
+                     FLOAT[:, ::1] XY,
                      FLOAT[:, ::1] V,
                      float R,
                      int i, int count,
                      UINT[:, ::1] Grid,
-                     float grid_spacing) nogil:
+                     float grid_spacing,
+                     omp_lock_t *locks
+                     ) nogil:
     cdef:
         FLOAT *XY1, *XY2, *V1, *V2
         int j, dim
         float eps = 1e-5
-
+        int grid_x, grid_y, grid_size 
+        int g_x,g_y
     # SUBPROBLEM 4: Add locking
     XY1 = &(XY[i, 0])
     V1 = &(V[i, 0])
+    acquire(&locks[i])
     #############################################################
     # IMPORTANT: do not collide two balls twice.
     ############################################################
     # SUBPROBLEM 2: use the grid values to reduce the number of other
     # objects to check for collisions.
-    for j in range(i + 1, count):
-        XY2 = &(XY[j, 0])
-        V2 = &(V[j, 0])
-        if overlapping(XY1, XY2, R):
-            # SUBPROBLEM 4: Add locking
-            if not moving_apart(XY1, V1, XY2, V2):
-                collide(XY1, V1, XY2, V2)
-
-            # give a slight impulse to help separate them
-            for dim in range(2):
-                V2[dim] += eps * (XY2[dim] - XY1[dim])
 
+    grid_x = int(XY[i,0] / grid_spacing)
+    grid_y = int(XY[i,1] / grid_spacing)
+    grid_size = Grid.shape[0]
+
+    for g_x in range(grid_x+1, grid_x+4):
+        for g_y in range(grid_y+1, grid_y+4):
+            if g_x>0 and g_x<grid_size and g_y>0 and g_y<grid_size and Grid[g_x,g_y]<count:
+                acquire(&locks[Grid[g_x, g_y]])
```

Comment:
Are you locking only one element here, or two elements (the ball currently being looked at and the ball that it overlaps with)?